### PR TITLE
feat(#8821): track Apdex performance

### PR DESCRIPTION
**Pull request by latin-panda**
# Description

Adds service to track performance 
Records Apdex in Telemetry (per user):

**Overall boot time:** including loading the code, purging, and accessing the database
  - boot_time
  - boot_time:apdex:[satisfied|tolerable|frustrated]

**Contacts tab:**
  - contact list
    - contact_list:load
    - contact_list:load:apdex:[satisfied|tolerable|frustrated]
  - contact detail - i.e. when selecting a contact.
    - contact_detail:<contact_type>:load
       - contact_detail:<_form>:load:contact_data
       - contact_detail:<_form>:load:load_descendants
       - contact_detail:<_form>:load:load_reports
       - contact_detail:<_form>:load:load_targets
       - contact_detail:<_form>:load:load_tasks
       - contact_detail:<_form>:load:load_contact_summary
    - contact_detail:<contact_type>:load:apdex:[satisfied|tolerable|frustrated]
  - contact form: this are app forms and contact forms loaded in contacts tab
    - enketo:contacts:<_form>:<_action>:render  
    - enketo:contacts:<_form>:<_action>:render:apdex:[satisfied|tolerable|frustrated]  
    - enketo:contacts:<_form>:<_action>:save  
    - enketo:contacts:<_form>:<_action>:save:apdex:[satisfied|tolerable|frustrated]  
  - search and/or sort (internally the sort is like a filter) 
    - contact_list:query
    - contact_list:query:apdex:[satisfied|tolerable|frustrated]

**Reports tab:**
  - report list
    - report_list:load
    - report_list:load:apdex:[satisfied|tolerable|frustrated]
  - select a report  
    - report_detail:<_form>:load
    - report_detail:<_form>:load:apdex:[satisfied|tolerable|frustrated]
  - report form
    - enketo:reports:<_form>:<_action>:render  
    - enketo:reports:<_form>:<_action>:render:apdex:[satisfied|tolerable|frustrated]  
    - enketo:reports:<_form>:<_action>:save  
    - enketo:reports:<_form>:<_action>:save:apdex:[satisfied|tolerable|frustrated]  
  - search and/or filter
    - report_list:query
    - report_list:query:apdex:[satisfied|tolerable|frustrated]

**Tasks tab:**
  - task list (telemetry should be the same as original)
    - tasks:load 
    - tasks:load:apdex:[satisfied|tolerable|frustrated]
  - task form
    - enketo:tasks:<_form>:<_action>:render  
    - enketo:tasks:<_form>:<_action>:render:apdex:[satisfied|tolerable|frustrated]  
    - enketo:tasks:<_form>:<_action>:save  
    - enketo:tasks:<_form>:<_action>:save:apdex:[satisfied|tolerable|frustrated]  

**Targets tab:**
  - target list
    - analytics:targets:load
    - analytics:targets:load:apdex:[satisfied|tolerable|frustrated]
  - target aggregates
    - analytics:target_aggregates:load
    - analytics:target_aggregates:load:apdex:[satisfied|tolerable|frustrated]

**Message tab:**
  - message list
    - message_list:load
    - message_list:load:apdex:[satisfied|tolerable|frustrated]
 - select conversation
   - messages_detail:load
   - messages_detail:load:apdex:[satisfied|tolerable|frustrated]

[CHT Docs PR](https://github.com/medic/cht-docs/pull/1278)

See telemetry mapping [here](https://docs.google.com/document/d/1iXZybHL56l3Z7Y_aEpRqkZkQYggG97EtpnawzA8N_zA/edit#heading=h.azy63xxyzh10)
#8821

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8821-track-apdex-performance/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8821-track-apdex-performance/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8821-track-apdex-performance/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: webapp/src/ts/modules/tasks/tasks.component.ts
Comment (latin-panda) [inline webapp/src/ts/modules/tasks/tasks.component.ts:164]: @garethbowen, quick question: how important is keeping the old telemetry names? 
We intend to prefix performance records with `perf:<component>:<etc>`, but thinking that we need backward compatibility when adding the ApDex, I added a flag to skip the prefix. 
It's a small thing, but having the name convention across would be nicer and avoid this extra check.

File: webapp/src/ts/services/performance.service.ts
Comment (garethbowen) [inline webapp/src/ts/services/performance.service.ts:91]: This is repetitive and creates a new data structure that's unnecessary. Refactor it so it just returns the satisfaction label, eg:

```suggestion
  private getSatisfactionLabel(time: number) {

    if (time <= this.APDEX_T) {
      return this.APDEX_SATISFIED;
    }
    
    if (time <= (this.APDEX_TOLERANCE * this.APDEX_T)) {
      return this.APDEX_TOLERATING;
    }
    
    return APDEX_FRUSTRATED;
  }
```

File: webapp/src/ts/services/performance.service.ts
Comment (garethbowen) [inline webapp/src/ts/services/performance.service.ts:65]: I don't think we should duplicate this because it doubles the number of records that need to be stored and increases the docs to sync. We can easily aggregate them on the server if we record the `component`.

File: webapp/src/ts/services/performance.service.ts
Comment (garethbowen) [inline webapp/src/ts/services/performance.service.ts:40]: Having multiple booleans in an API is annoying because when you're reading the code you have no idea what `stop(true, false)` means and have to dig down deeper.

Consider changing this to something like...

```suggestion
      stop: ({ recordApdex = false, prefix = true }) => this.recordPerformance(name, startTime, recordApdex, prefix),
```

File: webapp/src/ts/services/performance.service.ts
Comment (garethbowen) [inline webapp/src/ts/services/performance.service.ts:14]: I find it complicated when the separator is part of the value here because if you want to add or change the telemetry name it's easy to put the `:` in the wrong place.

My suggestion is to remove the `:` and instead of concatenating the string use join, eg:

```
[this.PERFORMANCE_PREFIX, 'app', this.APDEX_SUBFIX, this.APDEX_AGGREGATE].join(':');
```

Comment (garethbowen) [inline webapp/src/ts/services/performance.service.ts:14]: Also, should this be SUBFIX or SUFFIX? A "subfix" is something else...

File: webapp/src/ts/modules/tasks/tasks.component.ts
Comment (garethbowen) [reply webapp/src/ts/modules/tasks/tasks.component.ts:164]: I think we have to keep backwards compatibility here - it's part of the API and someone may be using it.

File: webapp/src/ts/effects/contacts.effects.ts
Comment (garethbowen) [inline webapp/src/ts/effects/contacts.effects.ts:60]: Move this below the `if(!id)` check - there's no point starting the track and then returning before you stop it.

File: webapp/src/ts/services/performance.service.ts
Comment (garethbowen) [inline webapp/src/ts/services/performance.service.ts:59]: It'd be clearer to me if this were called `duration` not `time`

File: webapp/src/ts/effects/contacts.effects.ts
Comment (garethbowen) [inline webapp/src/ts/effects/contacts.effects.ts:94]: It's really weird to set the name on creation and on stop. Let's change the API so we only do it on stop? eg:

```suggestion
            trackPerformance?.stop(this.selectContactTrack, true);
```

Then I think you can clean up the code even more if instead of storing the name in `this.selectContactTrack` you calculate it only when you need it in the `finally` block.

**Review by garethbowen — CHANGES_REQUESTED** [review]
Can't wait to see the apdex data flowing in!

Some suggestions inline.

File: webapp/src/ts/services/performance.service.ts
Comment (latin-panda) [reply webapp/src/ts/services/performance.service.ts:65]: > it doubles the number of records that need to be stored and increases the docs to sync

There was a back-and-forth on the aggregate, but it can also be calculated when building the reports

**Comment by latin-panda** [comment]
@ralfudx @Benmuiruri @garethbowen This is ready, please review, thanks!

**Comment by latin-panda** [comment]
Testing evidence

<details> 
<summary>Messages tab</summary>

select-conversation-tolerable
select-conversation-frustrated
select-conversation-satisfied
message-list-satisfied
</details>

<details> 
<summary>Reports tab</summary>

report-list-all-apedex
select-report-frustrated
select-report-tolerable
select-report-satisfied
report-list-satisfied
report-form-satisfied
</details>

<details> 
<summary>Contacts tab</summary>

contact-form-satisfied
contact-report-forms-edit-add
select-contact-satisfied
select-contact-all
contact-list-satisfied
</details>


<details> 
<summary>Targets tab</summary>

analytics-satisfied
</details>

File: webapp/src/ts/effects/contacts.effects.ts
Comment (garethbowen) [inline webapp/src/ts/effects/contacts.effects.ts:97]: Would it be easier to get the selectedContact from the store rather than in a variable?

**Review by garethbowen — COMMENTED** [review]
I've had a quick look and left a comment - was there something specific you wanted me to look at?

I'll leave it to the other reviewers to do a complete review.

File: webapp/src/ts/effects/contacts.effects.ts
Comment (latin-panda) [reply webapp/src/ts/effects/contacts.effects.ts:97]: I'm afraid of a race condition or retriggering this effect and causing more spaghetti in the execution flow. 
This is not the nicest, but it's safe and simple to debug. Ideally, we refactor all this code to simplify, but not in this PR scope.

**Review by Benmuiruri — APPROVED** [review]
Impressive work Jenni. It's great to see those logs.

**Comment by michaelkohn** [comment]
This is looking cool! 

I think the initial comment that says `contact form: this are app forms and contact forms loaded in contacts tab` is a bit confusing, but the data samples above look the way I'd expect them. 

A question about `enketo` related telemetry... here's what I believe we have today:

### Contacts Tab
- `enketo:contacts:<form>:<action>:<component>` for a Contact form
- `enketo:reports:<form>:<action>:<component>` for an App form (report)

### Reports Tab
- `enketo:reports:<form>:action>:<component>` for an App form (report)
_It's not possible to access a Contact form from the Reports tab_

### Tasks Tab
- `enketo:tasks:<form>:action>:<component>` for a Task form (report)
_It's not possible to access a Contact form from the Tasks tab_

It seems like we're kind of inconsistent in that we **don't differentiate** Enketo reports accessed from Reports vs Contacts tabs, but we **do** differentiate Enketo reports accessed from the Tasks tab. I know this is not something new (ie it's how the system has always worked), but wondering if I'm misunderstanding this or if there is a specific reason why it's the same for Contacts/Reports but different for Tasks. Is there a safe way to change it so that we can differentiate between reports accessed from Contacts vs. Reports?

**Comment by latin-panda** [comment]
Hi @ralfudx do you have any feedback?

**Comment by ralfudx** [comment]
Yes @latin-panda this looks great
However, I'm curious to know how we intend to capture the report (especially for multiple users) - In the Testing evidence you attached, i can see min, max, and the sum but the `average response time` which ideally should be a factor of the `count` is not represented
Maybe we can have a chat about this during the teams call?

**Review by ralfudx — APPROVED** [review]

**Comment by latin-panda** [comment]
> the Testing evidence you attached, i can see min, max, and the sum but the average response time which ideally should be a factor of the count is not represented

Good question. Telemetry doesn't calculate an average, [see docs](https://docs.communityhealthtoolkit.org/apps/guides/performance/telemetry/). But it has a `sumsqr`, if I remember @michaelkohn's explanation correctly, it's the variance in the occurrences.

The telemetry are saved in the user's meta database in CouchDB, you'll find a telemetry document per day that the CHW was using the app.

File: webapp/src/ts/modules/contacts/contacts.component.ts
Comment (latin-panda) [inline webapp/src/ts/modules/contacts/contacts.component.ts:412]: This differs from the telemetry record `search:contacts:<filter[:filter]>`. 
The `contact_list:query` tracks the query in the Contacts Tab only and includes the whole query process to display the result, like formatting contacts, setting up the scroll etc. This one also tracks the Apdex. 
While `search:contacts:<filter[:filter]>` tracks the process from search service only, which can be called by other components (not just Contact list).

**Comment by latin-panda** [comment]
@michaelkohn @Benmuiruri I have added the Apdex to the search in the Contacts and Reports Tab, as agreed yesterday during the team call.

**Comment by latin-panda** [comment]
Mentioning here a bug that @Benmuiruri resolved [here](https://github.com/medic/cht-core/pull/8838) but still needs confirmation from CT before merging

File: webapp/tests/karma/ts/modules/messages/messages-content.component.spec.ts
Comment (Benmuiruri) [inline webapp/tests/karma/ts/modules/messages/messages-content.component.spec.ts:153]: Do we keep this as plural or make it singular?  The other entries are singular i.e `report_detail`, `contact_detail`

Comment (latin-panda) [reply webapp/tests/karma/ts/modules/messages/messages-content.component.spec.ts:153]: Plural, see conversation [here](https://medic.slack.com/archives/C02GME3BQE6/p1707191934904099?thread_ts=1707166546.743469&cid=C02GME3BQE6). Because when clicking on one item from the list, it loads a conversation containing many messages. 
Different from reports that load 1 report, for example.

**Comment by latin-panda** [comment]
@Benmuiruri @ralfudx please review one last time before we merge this work

**Review by Benmuiruri — APPROVED** [review]
LGTM 🚀

**Review by ralfudx — APPROVED** [review]
LGTM as well

**Pull request merged**
